### PR TITLE
feat: allow to set proto material to portals of volumes and containers

### PR DESCRIPTION
**Pull request by asalzburger**
This PR prepares for new Grid bases surface material by making the `ProtoSurfaceMaterial` either dependent on `BinUtility` or `ProtoBinning` via a simple template specialisation.

It then allows `ProtoGridMaterial` to be attached to portal surfaces of `Volume` objects and `surface` objects.

It adds also unit tests to those.

### CODE REVIEW

**Comment by asalzburger** [comment]
Hey @Corentin-Allaire - I think this is best reviewed by you.

**Comment by Corentin-Allaire** [comment]
> Hey @Corentin-Allaire - I think this is best reviewed by you.

@asalzburger, I am in a ML conference this week I will review it first thing next week

**Comment by asalzburger** [comment]
Hey @Corentin-Allaire - this is ready to review now.

File: Core/include/Acts/Material/ProtoSurfaceMaterial.hpp [hunk 1]
```diff
@@ -20,112 +20,108 @@ namespace Acts {
...
+/// binning description
 ///
 /// The ProtoSurfaceMaterial class acts as a proxy to the SurfaceMaterial
 /// to mark the layers and surfaces on which the material should be mapped on
 /// at construction time of the geometry and to hand over the granularity of
 /// of the material map with the bin Utility.
-
-class ProtoSurfaceMaterial : public ISurfaceMaterial {
+template <typename BinningType>
+class ProtoSurfaceMaterialT : public ISurfaceMaterial {
```
Comment (Corentin-Allaire) [inline Core/include/Acts/Material/ProtoSurfaceMaterial.hpp:31]: What does the T means here ?

File: Core/include/Acts/Material/ProtoSurfaceMaterial.hpp [hunk 2]
```diff
@@ -20,112 +20,108 @@ namespace Acts {
...
-  /// Constructor without BinUtility - homogeneous material
-  ProtoSurfaceMaterial() = default;
+  /// Constructor without binningType - homogeneous material
+  ProtoSurfaceMaterialT() = default;
 
-  /// Constructor with BinUtility - multidimensional material
-  ///
-  /// @param binUtility a BinUtility determining the granularity
-  ///        and binning of the material on the surface/layer
+  /// Constructor with BinnintType
```
Comment (Corentin-Allaire) [inline Core/include/Acts/Material/ProtoSurfaceMaterial.hpp:36]:
```suggestion
  /// Constructor with BinningType
```

**Review by Corentin-Allaire — APPROVED** [review]
Looks good to me I have just one small question. But this can go in

File: Core/include/Acts/Material/ProtoSurfaceMaterial.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Material/ProtoSurfaceMaterial.hpp:31]: Indicate that it's the template holder.

**Comment by asalzburger** [comment]
I rebased this one.

File: Core/include/Acts/Material/ProtoSurfaceMaterial.hpp (hunk 1 above)
Comment (Corentin-Allaire) [reply Core/include/Acts/Material/ProtoSurfaceMaterial.hpp:31]: Ah make sense !

**Review by Corentin-Allaire — DISMISSED** [review]

File: Core/include/Acts/Detector/detail/ProtoMaterialHelper.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+struct BinningDescription;
+
+namespace detail {
+namespace ProtoMaterialHelper {
+
+/// @brief Method that attaches proto material to a surface given
+/// a proto binning description
+///
+/// @param gctx is the geometry context, needed for extent measureing
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Detector/detail/ProtoMaterialHelper.hpp]: fixed.

**Comment by asalzburger** [comment]
Hi @Corentin-Allaire  - can you re-approve, I had to resolve some merge conflicts.

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [66908fc54935a812070e52ec386202dd8ccbe6c3]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779252)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779251)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779250)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779249)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779248)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779247)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779246)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779245)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34779241)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778973)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778972)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778971)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778969)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778968)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778967)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778964)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778961)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778960)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778959)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778956)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778954)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778952)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34778951)||

**Pull request merged**